packages/frontend/src/contexts/Web3Context.tsx

```diff
@@ -261,64 +269,73 @@ const Web3ContextProvider: FC = ({ children }) => {
+      logger.debug('checkConnectedNetworkId', networkId, signerNetworkId)
+
+      if (networkId.toString() === signerNetworkId?.toString()) {
+        return true
+      }
+
       onboard.config({ networkId })
       if (onboard.getState().address) {
-        let nativeCurrency: any
-        if (networkIdToSlug(networkId) === 'gnosis') {
-          nativeCurrency = {
-            name: 'xDAI',
-            symbol: 'XDAI',
-            decimals: 18,
-          }
-        } else if (networkIdToSlug(networkId) === 'polygon') {
-          nativeCurrency = {
-            name: 'Matic',
-            symbol: 'MATIC',
-            decimals: 18,
-          }
-        }
-
         try {
-          if (provider && networkId) {
-            const wantNetworkName = networkNames[networkId] || 'local'
-            const isL1 = ['Mainnet', 'Ropsten', 'Rinkeby', 'Goerli', 'Kovan'].includes(
-              wantNetworkName
-            )
+          const wantNetworkName = networkNames[networkId] || 'local'
+          const isL1 = ['Mainnet', 'Ropsten', 'Rinkeby', 'Goerli', 'Kovan'].includes(
+            wantNetworkName
+          )
+
+          if (isL1) {
+            await provider?.send('wallet_switchEthereumChain', [
+              {
+                chainId: `0x${Number(networkId).toString(16)}`,
+              },
+            ])
+          } else {
+            let nativeCurrency: any
+
+            if (networkId === ChainId.Gnosis) {
+              nativeCurrency = {
+                name: 'xDAI',
+                symbol: 'XDAI',
+                decimals: 18,
+              }
+            } else if (networkId === ChainId.Polygon) {
+              nativeCurrency = {
+                name: 'Matic',
+                symbol: 'MATIC',
+                decimals: 18,
+              }
+            }
+
             const rpcObj = {
               chainId: `0x${Number(networkId).toString(16)}`,
-              chainName: networkNames[networkId.toString()],
+              chainName: networkNames[networkId],
               rpcUrls: [getRpcUrl(networkIdToSlug(networkId.toString()))],
               blockExplorerUrls: [getBaseExplorerUrl(networkIdToSlug(networkId.toString()))],
               nativeCurrency,
             }
-            if (isL1) {
-              await provider?.send('wallet_switchEthereumChain', [
-                {
-                  chainId: `0x${Number(networkId).toString(16)}`,
-                },
-              ])
-            } else {
-              await provider?.send('wallet_addEthereumChain', [rpcObj])
-            }
-          }
 
-          return true
+            await provider?.send('wallet_addEthereumChain', [rpcObj])
+          }
         } catch (err) {
           logger.error(err)
         }
-
-        await onboard.walletCheck()
       }
-      return false
-    }
+      const p = await provider.getNetwork()
+      if (p.chainId === networkId) {
+        return true
+      }
```

Comment:
previously, this function would `return true` simply after sending the `wallet_addEthereumChain`, regardless of what the user selected. now we double-check to make sure the network matches.